File: dynabuffers-java/src/main/kotlin/dynabuffers/DynabuffersEngine.kt
package dynabuffers

import dynabuffers.api.IAnnotation
import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import dynabuffers.api.map.DynabuffersMap
import dynabuffers.ast.*
import dynabuffers.ast.annotation.*
import dynabuffers.ast.structural.Annotation
import dynabuffers.exception.DynabuffersException
import dynabuffers.header.RootElement
import java.nio.ByteBuffer

class DynabuffersEngine(private val tree: List<IType>) {

    private val listeners = ArrayList<(String) -> Unit>()

    fun addListener(consumer: (String) -> Unit) = listeners.add(consumer)

    fun serialize(map: Map<String, Any?>): ByteArray {
        val root = RootElement(tree)
        val registry = Registry(tree, listeners)
        val buffer = ByteBuffer.allocate(root.size(map, registry))
        root.serialize(map, buffer, registry)
        return buffer.array()
    }

    fun serialize(map: Map<String, Any?>, namespaceName: String): ByteArray {
        if (map.containsKey(SpecialKey.NAMESPACE.key)) throw IllegalStateException("Nope sir")
        return serialize(map.plus(SpecialKey.NAMESPACE.key to namespaceName))
    }

    fun serialize(map: Map<String, Any?>, namespaceNames: List<String>): ByteArray {
        if (map.containsKey(SpecialKey.NAMESPACE.key)) throw IllegalStateException("Nope sir")
        return serialize(map.plus(SpecialKey.NAMESPACE.key to namespaceNames.joinToString(".")))
    }

    fun serialize(result: String) = serialize(mapOf("value" to result))
    fun serialize(result: ByteArray) = serialize(mapOf("value" to result))
    fun serialize(result: Int) = serialize(mapOf("value" to result))
    fun serialize(result: Long) = serialize(mapOf("value" to result))
    fun serialize(result: Short) = serialize(mapOf("value" to result))
    fun serialize(result: Byte) = serialize(mapOf("value" to result))
    fun serialize(result: Float) = serialize(mapOf("value" to result))
    fun serialize(result: Double) = serialize(mapOf("value" to result))
    fun serialize(result: Boolean) = serialize(mapOf("value" to result))


    fun deserialize(bytes: ByteArray): DynabuffersMap {
        val bb = ByteBuffer.wrap(bytes)
        return RootElement(tree).deserialize(bb, Registry(tree, listeners))
    }
}

File: dynabuffers-python/dynabuffers/DynabuffersEngine.py
from typing import Union, Any, List

from dynabuffers import NAMESPACE_KEY
from dynabuffers.NamespaceResolver import NamespaceResolver
from dynabuffers.Registry import Registry
from dynabuffers.api.ISerializable import ISerializable, ByteBuffer
from dynabuffers.api.map.DynabuffersMap import DynabuffersMap
from dynabuffers.header.RootElement import RootElement


class DynabuffersEngine(object):

    def __init__(self, tree: [ISerializable]):
        self.tree = tree
        self.namespace_resolver = NamespaceResolver(tree)
        self.listeners = []

    def add_listener(self, listener):
        self.listeners.append(listener)

    def serialize(self, value: Union[dict, Any], namespace_names: Union[List[str], str] = None) -> bytearray:
        if not isinstance(value, dict):
            return self.serialize({"value": value})

        if namespace_names is not None:
            if NAMESPACE_KEY in value:
                raise Exception("{} is already present in dictionary - use serialize/1".format(NAMESPACE_KEY))
            if isinstance(namespace_names, List):
                value[NAMESPACE_KEY] = '.'.join(namespace_names)
            if isinstance(namespace_names, str):
                value[NAMESPACE_KEY] = namespace_names

        root_element = RootElement(self.tree)
        registry = Registry(self.tree, self.listeners)
        buffer = ByteBuffer(root_element.size(value, registry))
        root_element.serialize(value, buffer, registry)
        return buffer.toBytes()

    def deserialize(self, bytes: bytearray, namespaces: Union[List[str], str] = None) -> DynabuffersMap:
        """
        :param bytes:
        :param namespaces: only there for compat reasons => unused and deprecated!
        :return:
        """
        bb = ByteBuffer(len(bytes), bytes)
        registry = Registry(self.tree, self.listeners)
        return RootElement(self.tree).deserialize(bb, registry)
